hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionStates.java

```diff
@@ -783,10 +801,20 @@ public void regionOffline(
             LOG.info("Found region in " + state +
               " to be reassigned by ServerCrashProcedure for " + sn);
             rits.add(hri);
-          } else if(state.isSplittingNew() || state.isMergingNew()) {
-            LOG.info("Offline/Cleanup region if no meta entry exists, hri: " + hri +
-                " state: " + state);
-            regionsToClean.add(state.getRegion());
+          } else if (state.isSplittingNew() || state.isMergingNew()) {
+            LOG.info(
+              "Offline/Cleanup region if no meta entry exists, hri: " + hri + " state: " + state);
+            if (daughter2Parent.containsKey(hri.getEncodedName())) {
+              HRegionInfo parent = daughter2Parent.get(hri.getEncodedName());
+              HRegionInfo info = getHRIFromMeta(parent);
+              if (info != null && info.isSplit() && info.isOffline()) {
+                regionsToClean.add(Pair.newPair(state.getRegion(), info));
```

Comment:
Oh, so parent is **OFFLINE** and **SPLIT=true** in meta only, but not in _assignedRegions_? Sounds wrong that we marked parent as OFFLINE and SPLIT even when splitting was still not entirely complete.

Reply:
If the RS machine crash when the SplitTransactionImpl step after PONR, master will handle the split rollback. Under normal conditions, it cleanup daughter region and try to assign parent region. If assign failed, parent region RIT so CatalogJanitor blocked; but at this time the master switch, new master will delete parent region because of **OFFLINE and SPLIT=true**.

Reply:
This daughter2Parent added stingency makes sense to me. Nice.

Reply:
> If the RS machine crash when the SplitTransactionImpl step after PONR, master will handle the split rollback.

Yeah, but it seems to work only because Master has different info for the region than what is actual in meta. What if Active Master also crashes before it triggers SCP and daughters get cleaned up? The other master will load regions state as seen in meta, then it can run into same problem again.

Reply:
> > If the RS machine crash when the SplitTransactionImpl step after PONR, master will handle the split rollback.
> 
> Yeah, but it seems to work only because Master has different info for the region than what is actual in meta. What if Active Master also crashes before it triggers SCP and daughters get cleaned up? The other master will load regions state as seen in meta, then it can run into same problem again.

daughter cleanup is excuted in SCP, and in this patch it will do three things together:
1.delete daughter region in meta
2.update parent region in meta
3.delete daughter region dir in HDFS
`	          if (regionPair != null) {
            MetaTableAccessor.deleteRegion(this.server.getConnection(), hri);	            MetaTableAccessor.deleteRegion(this.server.getConnection(), hri);
          }	          }
          if (parentInfo != null) {
            List<Mutation> mutations = new ArrayList<Mutation>();
            HRegionInfo copyOfParent = new HRegionInfo(parentInfo);
            copyOfParent.setOffline(false);
            copyOfParent.setSplit(false);
            Put putParent = MetaTableAccessor.makePutFromRegionInfo(copyOfParent);
            mutations.add(putParent);
            MetaTableAccessor.mutateMetaTable(this.server.getConnection(), mutations);
          }
          LOG.debug("Cleaning up HDFS since no meta entry exists, hri: " + hri);	          LOG.debug("Cleaning up HDFS since no meta entry exists, hri: " + hri);
          FSUtils.deleteRegionDir(server.getConfiguration(), hri);	          FSUtils.deleteRegionDir(server.getConfiguration(), hri);`

So if Active Master also crashes before it triggers SCP, the daughter won`t be deleted.

Reply:
> So if Active Master also crashes before it triggers SCP, the daughter won`t be deleted.

Yes, that's my point. We would now have a new active master that sees parent split as complete, although split was midway through. It will potentially remove the parent, and fail to online the daughters. 

I wonder if working on correcting the regions state and split flag updates would sort split failures at different scenarios. It also does not seem consistent the way we do these updates in meta only and don't reflect it on the "in-memory" region info master has.